File: hilti/python/hilti/codegen/caddr.py
# $Id$
#
# Code generator for caddr instructions.

import llvm.core

from hilti.core import *
from hilti import instructions
from codegen import codegen

_doc_c_conversion = """ A ``caddr`` is mapped to ``void *``."""

@codegen.typeInfo(type.CAddr)
def _(type):
    typeinfo = codegen.TypeInfo(type)
    typeinfo.c_prototype = "void *"
    typeinfo.to_string = "hlt::caddr_to_string";
    return typeinfo

@codegen.llvmDefaultValue(type.CAddr)
def _(type):
    return llvm.core.Constant.null(codegen.llvmTypeGenericPointer())

@codegen.llvmType(type.CAddr)
def _(type, refine_to):
    return codegen.llvmTypeGenericPointer()

@codegen.when(instructions.caddr.Function)
def _(self, i):
    
    fid = i.op1().value()
    func = self.lookupFunction(fid)
    builder = self.builder()

    if func.callingConvention() == function.CallingConvention.HILTI:
        (main, resume) = self.llvmCStubs(func)
        main = builder.bitcast(main, self.llvmTypeGenericPointer())
        resume = builder.bitcast(resume, self.llvmTypeGenericPointer())
    else:
        util.internal_error("caddr.Function not supported for non-HILTI functions yet")
        
    struct = llvm.core.Constant.struct([main, resume])
    codegen.llvmStoreInTarget(i.target(), struct)

File: hilti/codegen/instructions/old/caddr.cc

#include <hilti.h>

#include "../stmt-builder.h"

using namespace hilti;
using namespace codegen;


void StatementBuilder::visit(statement::instruction::caddr::Function* i)
{

/*
    auto op1 = cg()->llvmValue(i->op1(), X);

    auto result = builder()->

    cg()->llvmStore(i, result);
*/

/*
    CodeGen::expr_list args;
    args.push_back(i->op1());

    auto result = cg()->llvmCall("hlt::X", args);

    cg()->llvmStore(i, result);
*/


/*
    def _codegen(self, cg):

        fid = self.op1().value()
        func = cg.lookupFunction(fid)
        builder = cg.builder()

        if func.callingConvention() == function.CallingConvention.HILTI:
            (hltmain, main, hltresume, resume) = cg.llvmCStubs(func)
            main = builder.bitcast(main, cg.llvmTypeGenericPointer())
            resume = builder.bitcast(resume, cg.llvmTypeGenericPointer())
        elif func.callingConvention() == function.CallingConvention.C_HILTI:
            func = cg.llvmFunction(func)
            main = builder.bitcast(func, cg.llvmTypeGenericPointer())
            resume = llvm.core.Constant.null(cg.llvmTypeGenericPointer())
        else:
            util.internal_error("caddr.Function not supported for non-HILTI/HILTI-C functions yet")

        struct = llvm.core.Constant.struct([main, resume])
        cg.llvmStoreInTarget(self, struct)

*/
}

